tests/framework/microvm.py

```diff
@@ -219,6 +225,47 @@ def kill(self):
                 self.memory_monitor.join(timeout=1)
             self.memory_monitor.check_samples()
 
+    def _validate_api_response_times(self):
+        """
+        Parses the firecracker logs for information regarding api server request processing times, and asserts they
+        are within acceptable bounds.
+        """
+        # Log messages are either
+        # 2023-06-16T07:45:41.767987318 [fc44b23e-ce47-4635-9549-5779a6bd9cee:fc_api] The API server received a Get request on "/mmds".
+        # or
+        # 2023-06-16T07:47:31.204704732 [2f2427c7-e4de-4226-90e6-e3556402be84:fc_api] The API server received a Put request on "/actions" with body "{\"action_type\": \"InstanceStart\"}".
+        api_request_regex = re.compile(
+            r":fc_api\] The API server received a (?P<method>\w+) request on \"(?P<url>(/\w*)+)\"( with body (?P<body>.*))?\."
+        )
+        api_request_times_regex = re.compile(
+            r":fc_api\] Total previous API call duration: (?P<execution_time>\d+) us.$"
+        )
+
+        # Note: Processing of api requests is synchronous, so these messages cannot be torn by concurrency effects
+        log_lines = self.log_data.split("\n")
+
+        current_method = None
+        current_url = None
+        current_body = None
+
+        for log_line in log_lines:
+            print(log_line)
+            match = api_request_regex.search(log_line)
+
+            if match:
+                current_method = match.group("method")
+                current_url = match.group("url")
+                current_body = match.group("body")
+
+            match = api_request_times_regex.search(log_line)
+
+            if match and current_url != "/snapshot/create":
+                exec_time = float(match.group("execution_time")) / 1000.0
+
+                assert (
+                    exec_time <= MAX_API_CALL_DURATION_MS
+                ), f"{current_method} {current_url} API call exceeded maximum duration: {exec_time} ms. Body: {current_body}"
+
```

Comment:
It would be good to check that we always get 1st and 2nd print in sequence so I would suggest clearing current_url here and have below 2 checks:
1.
```
if match:
    if current_url is not None:
        current_url = match.group("url")
    else:
        assert "something like got 1st pattern of prints twice"
```
2.
```
match = api_request_times_regex.search(log_line)
if match and current_url is None:
    assert "got 2nd pattern print without the 1st, so time did exceed but we don't know for which api since print having api name is missing from logs"
```

Reply:
sure